Refactor project types & add Next.js

AWolf81 opened the pull request:
**Related Issue:**
#59

**Summary:**
This branch is based on branch `nextjs-flow` (flow for workflow). I first started with this and I pushed it just for completeness. After merging the refactor branch we can safely delete the other branch.
I've merged the `process-logging` branch into this branch so it was easier to develop the refactoring.

- Created a configuration file inside `src/config/project-types.js` which contains three objects for storing the following information:
  - devServer task name
  - devServer command args & port handling with "template" variable
- Modified `isDevServerTask`, `getDevServerTaskForProjectId` & `getTaskType` to use the configuration object.

As mentioned in the issue comment I first had the idea of having a function inside of the configuration object but I don't like that.
Now I'm using the string `$port` inside of the configuration that will be replaced with the real port number in code. Just top-level replacement of `$port` is not working but I think we could add this later. Also not sure if we need to support deeply nesting as it is only working for one level e.g.
```js
{
  env: {
    PORT: '$port'
  }
}
```

Will be replaced to (if available port is 3000):
```js
{
  env: {
    PORT: 3000
  }
}
```
I think this is also the first step for improving the Gatsby workflow. Do we have an issue for Gatsby improvement? I think I read that on Gitter but I'm not sure what should be improved.
 
**Things to improve / discuss:**
- Nextjs logo is a bit small and I don't like that we're having Next.js text twice. Maybe we could remove the second text. (see screeshot below)
- Is the approach with the configuration in one file OK? I think that's better than the dynamic require. I'm also not sure if configuration is the right word for it. It's more like a mapping but I don't have a better word.
- I tried to create a common structure with the configuration object but there is at least one position that is still adding some noise. e.g. `task.saga` line 179 adds `additionalArgs` to the command if test task running. We could also add this to the configuration but I think we can modify this later.
- Should we swallow some log messages from Next.js? I think we don't need that many messages.

**Todo**
- [x] Check Gatsby Format & Test on master. It fails on this branch. This issue is not related to this branch. Same behaviour on master. Needs to be tackled separately.
- [x] I need to file an issue at `create-next-app` because with-out my [fork](https://github.com/AWolf81/create-next-app) it wasn't working with the projectDirectory passed to the script  `npx create-next-app c:\users\alexander\Guppy-dev\test` wasn't working. Maybe I can create a PR there if no other solution is available. OK, added a comment [here](https://github.com/segmentio/create-next-app/issues/30)
- [x] Check `Package.json` of Next.js project and why there is `start` task missing. There is a `start` task but it's not visible in Guppy as Melanie mentioned. **Checked start script** Seems like it requires to run build before. So `dev` is the right command to start the development server. `Start` is not displayed as we're only filtering by `task name` so it gets removed from tasks list. I'll update this so we're more specific e.g. `start` & `create-react-app`. OK, I've modified this so the filtering is working but there is another issue because `start` from next.js is also a long-running script which launches a local server we're having trouble with it. I'll open an issue for it so I can describe it in detail.
- [x] Add triangle next icon as Aaron proposed
- [x] Modify comment for isDevServer as proposed by j-f1

**Screenshots/GIFs:**
**Project create with Next.js**
![grafik](https://user-images.githubusercontent.com/3046542/47255699-84dae400-d475-11e8-9a93-c7261bcaffc2.png)


**Next.js project**
![grafik](https://user-images.githubusercontent.com/3046542/47255710-9c19d180-d475-11e8-8265-8485eaba90e3.png)
(failed build because I tested task stopping.)

AWolf81 commented on src/reducers/tasks.reducer.js:
```diff
@@ -251,7 +256,7 @@ export const getTaskDescription = (name: string) => {
 export const isDevServerTask = (name: string) =>
   // Gatsby and create-react-app use different names for the same task.
-  name === 'start' || name === 'develop';
+  devServerTaskNames.indexOf(name) !== -1;
```
Yes, you're right. `includes` is better as it's more intuitive. I'll change the comment & the `includes` later today.

melanieseltzer replied:
@j-f1 Even better :)

AWolf81 commented on src/sagas/task.saga.js:
```diff
@@ -349,27 +362,61 @@ const createStdioChannel = (
   });
 };
 
+export const substituteConfigVariables = (
```
I'll add an additional comment with more details before the function definition as I'd like to leave the comment in line 409 as I think it helps for the understanding there.

superhawk610 commented:
Looking around, I've seen this Next.js logo used a [couple](https://hackernoon.com/quickstart-a-next-js-website-using-a-headless-cms-api-7915887043c9) [places](https://github.com/cosmicjs/next.js-website), I think it would fit slightly better with the existing icons?

![next](https://user-images.githubusercontent.com/18172185/46974343-c80e0f00-d091-11e8-86cd-7729181baa25.png)

AWolf81 commented:
@melanieseltzer I think I've addressed your review changes. Could you please have a look?

**The only open things are:**
- Using my fork for create-next-app. I think we can change it later to the base repo once we're having a feedback from the maintainer (11 days with-out feedback).
- The second dev server `start` script is a problem as it's not detected as long running (see separate issue)
- Script description needs to be per project type (see separate issue)

I think we can tackle the open stuff afterwards.

melanieseltzer commented on src/sagas/task.saga.js:
```diff
@@ -349,27 +362,64 @@ const createStdioChannel = (
+          );
+        }
+      }
+      // todo: add top level substiution - not used yet but maybe needed later e.g. { env: $port } won't be replaced.
```
nit: typo (`substitution`)

melanieseltzer commented on src/config/project-types.js:
```diff
@@ -0,0 +1,54 @@
+      // not sure if we need that nesting but I think there could be more to configure
+      args: projectPath => [
+        // used for project creation previous getBuildInstructions
+        'github:awolf81/create-next-app', // later will be 'create-next-app' --> issue not filed yet
```
Perhaps update the comment for `'github:awolf81/create-next-app'` to reference the issue URL, for tracking?
